### ISSUE TITLE
Enhance Models Compare Function [ENHANCEMENT]

### ISSUE THREAD

**Issue by martinvonk** [pastas/pastas#428]
I would like to make a few changes to the pastas.plots.compare() function:

- [ ] Make it possible to compare the time series models for different locations. So the models don't need to have the same oseries per se. 
- [ ] Show the &#916;AIC instead of the R^2 if the models at the same location are shown. 
- [ ] Make the xlim of the respons function equal to the longest response instead of the first plotted response.
- [ ] Do something with the table since way too many are parameters (maybe give a list with parameters names to compare?)
- [ ] Add an option to normalize the contributions by subtracting the minimum value.
- [ ] Don't call model.plots.results() first but create a figure first and fill the axes.

Thoughts @dbrakenhoff & @raoulcollenteur ?

**Comment by dbrakenhoff** [comment]
All for! Or almost anyway:
- I'd still like to be able to see the R^2 I think, even for models at the same location (so maybe add a `stat` kwarg)? 
- What do you mean by "way to many are parameters" in the table? The current solution requires stressmodels to have the same names. Otherwise you'll need to specify which parameters you'd like to compare on each row, which is quite cumbersome and annoying? Plus then you need to decide if you show both parameter names. When i made this it felt like there wasn't really a good solution for this problem... So curious what your suggestion would be?
- Fully agree on scaling the step response plot better.
- Normalizing contributions sounds interesting (should be clear in the plot labels that this has been done). For pumping wells you'd probably prefer using the maximum value? For positive influences the minimum value and for influences that can be both positive and negative perhaps the median or mean?
- And good idea to replicate ml.plots.results() in the compare method, I think it'll be nicer to have everything contained in that method (even though it will be lots of code). 

Some other thoughts:
- A method to link differently named stressmodels so they appear in the same plot would be nice too. Maybe a dictionary with `{"<a general name for the plot label>": ["stressmodel_name1", "stresmodel_name2", "..."]}`. Once you decide to plot a stressmodel, that would require checking the other models whether they contain a model with one of those names and than add that to the resulting plots.
- Now the model with the most stressmodels is plotted first, but if two models contain different stresses, then one of those stresses is never shown on the compare plot. Perhaps determining the number of unique stressmodels (also taking into account the option above) would be useful, and then looping over the different models and plotting them in the right axes.
- The adjust_heights option makes it really difficult to scale axes properly and ensuring all stress contributions are visible. I believe if you turn off adjust_heights, the axes are rescaled to show all contributions. I like the adjust_heights option, especially for very similar models, but it can also be annoying. Something to keep in mind while rewriting the compare function...

**Comment by martinvonk** [comment]
Thank you for the extensive response and thinking :)

> * I'd still like to be able to see the R^2 I think, even for models at the same location (so maybe add a stat kwarg)?

I could also life if we show both. Both metrics are informative.

> * What do you mean by "way to many are parameters" in the table?

I find that the table generally scales very badly with the figure. So if there are too many parameters. The fontsize becomes too small and the text becomes unreadable. 

> * The current solution requires stressmodels to have the same names. Otherwise you'll need to specify which parameters you'd like to compare on each row, which is quite cumbersome and annoying? Plus then you need to decide if you show both parameter names. When i made this it felt like there wasn't really a good solution for this problem... So curious what your suggestion would be?

My solution would be either to:
1. Ditch the table entirely because we are visually comparing the models. If you want to compare the parameters use `ml.parameters`. 
2. Specify which parameters you want to compare for each model (or specify a substring e.g. '_A') 
3. Use a different library than matplotlib that visualizes tables nicely (e.g. plotly)

> * For pumping wells you'd probably prefer using the maximum value? For positive influences the minimum value and for influences that can be both positive and negative perhaps the median or mean?

I guess we can also determine whether the mean stress is positive or negative and then subtract the minimum or maximum value respectively.

> * A method to link differently named stressmodels so they appear in the same plot would be nice too. Maybe a dictionary with {"<a general name for the plot label>": ["stressmodel_name1", "stresmodel_name2", "..."]}. Once you decide to plot a stressmodel, that would require checking the other models whether they contain a model with one of those names and than add that to the resulting plots.

In pastastore there are kinds of stresses, maybe we can use that? 

> * Now the model with the most stressmodels is plotted first, but if two models contain different stresses, then one of those stresses is never shown on the compare plot. Perhaps determining the number of unique stressmodels (also taking into account the option above) would be useful, and then looping over the different models and plotting them in the right axes.

Agreed

> * The adjust_heights option makes it really difficult to scale axes properly and ensuring all stress contributions are visible. I believe if you turn off adjust_heights, the axes are rescaled to show all contributions. I like the adjust_heights option, especially for very similar models, but it can also be annoying. Something to keep in mind while rewriting the compare function...

Personally I would ditch the adjust_heights function because it makes life super complex.



Another random thought of mine:
I really like using the [matplotlib.pyplot.subpot_mosaic()](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.subplot_mosaic.html) function because it makes the adjusting layout of a figure super easy. We could use this to make the plotting highly flexible. So the users provides a custom mosaic, for instance:
```
[['oseries', 'oseries', 'parameters'],
 [f'contribution_{stressname1}', f'contribution_{stressname1}', f'response_{stressname1}']]
```
or a mosaic like:
```
[['oseries', 'oseries', 'metric_aic'],
 f'contribution_{stressname1}', f'contribution_{stressname1}',  f'parameters_{stressname1}']]
```
and we define internally what ax['oseries'], ax['parameters'] etc. look like and return the figure that belongs to that mosaic.

**Comment by dbrakenhoff** [comment]
Some psuedo code as an idea...

```python

class ModelComparison:
    def __init__(self, models, axes=None, combine=None):
        self.models = models
        if axes is None:
            # build axes using mosaic feature
            pass
        else:
            self.axes = axes

    def build_default_axes(self, n_stressmodels):
        """
        sim sim tab
        res res tab
        sm1 sm1 rf1
        sm2 sm2 rf2
        smn smn rfn
        """
        # think about x-axis sharing
        pass

    def get_unique_stressmodels(self, combine=None):
        # combine is dictionary w/ new name and stressmodels names to combine
        pass

    def plot_table(self, df, ax="table", **kwargs):
        pass

    def plot_contributions(self, names, ax="sm1", normalize=False, **kwargs):
        pass

    def plot_rfunc(self, ax="sm1_rf", block_or_step="step", **kwargs):
        pass

    def plot_oseries(self, models=None, ax="sim", **kwargs):
        pass

    def plot_simulation(self, ax="sim", **kwargs):
        pass

    def plot_residuals(self, ax="res", **kwargs):
        pass

    def plot_noise(self, ax="res", **kwargs):
        pass

    def plot_oseries_locations(self, ax="map", **kwargs):
        pass

    def plot_stresses_locations(self, ax="map", **kwargs):
        pass

    @staticmethod
    def _legend_helper(ax):
        pass
        # return handles, labels


# reconstructing model results plot
mc = ModelComparison([ml0, ml1])
# sim axis
mc.plot_oseries(
    models=[ml0.name]
)  # plot only 1 oseries, models are for same location
mc.plot_simulation()  # plot both simulations (think about how to control legend)
# residuals axis
mc.plot_residuals()  # plot both simulations (think about how to control legend)
mc.plot_noise()
# recharge axis
mc.plot_contributions(
    names=["recharge_lin", "recharge_nl"], ax="recharge", normalize=True
)
mc.plot_rfunc(names=["recharge_lin", "recharge_nl"], ax="recharge_rf")
# well axis
mc.plot_contributions(names="wells", ax="well")  # only one model has a well
mc.plot_rfunc(names="wells", ax="wells_rf")  # only one model has a well

# add table
df = pd.DataFrame(
    index=["ml0", "ml1"],
    columns=["evp", "wells_A"],
    data=[[71, np.nan], [85, -1.2]],
)
mc.plot_table(df)
```

**Comment by dbrakenhoff** [comment]
- I like the idea of passing in a dataframe to the plot_table, that gives the user all the control they should want, and maybe implement the current behavior in a plot_default_parameters_table method. Also if you don't want the table, you're free to leave it blank or plot a map. 
- Agreed on the mosaic too, I like the idea of named axes!
- Agreed on no adjust_heights in this implementation.

**Comment by martinvonk** [comment]
I created a notebook where I create a class like you mentioned @dbrakenhoff. 
https://github.com/pastas/pastas/blob/428-models-compare/examples/notebooks/22_comparing_models.ipynb
The class needs some improvement but I think we can work from here. 

I now create the class as such that you can also call dataframes with for instance the paramters of the different models (or a selection based on a list of (sub)strings). This way the class gets more functionalities, not only plotting.

**Comment by martinvonk** [comment]
Closed with #429
